File: byteback-test/test-java8/src/main/java/byteback/dummy/condition/Counter.java
package byteback.dummy.condition;

import static byteback.annotations.Contract.*;
import static byteback.annotations.Operator.*;
import static byteback.annotations.Special.*;

import byteback.annotations.Contract.Ensure;
import byteback.annotations.Contract.Predicate;

public class Counter {

	@Predicate
	public boolean increments_count_by_1() {
		return eq(count, old(count) + 1);
	}

	@Predicate
	public boolean increments_count_by_10() {
		return eq(count, old(count) + 10);
	}

	@Predicate
	public boolean count_is_0() {
		return eq(count, 0);
	}

	public static void main() {
		final Counter counter = new Counter();
		assertion(eq(counter.count, 0));
		counter.increment();
		assertion(eq(counter.count, 1));
		counter.countTo10();
		assertion(eq(counter.count, 11));
		counter.countTo10Indirectly();
		assertion(eq(counter.count, 21));
	}

	int count;

	@Ensure("count_is_0")
	public Counter() {
		this.count = 0;
	}

	@Ensure("increments_count_by_1")
	public void increment() {
		count++;
	}

	@Ensure("increments_count_by_10")
	public void countTo10() {
		final int old_count = count;

		for (int i = 0; i < 10; ++i) {
			invariant(lte(0, i) & lte(i, 10));
			invariant(eq(count, old_count + i));
			count++;
		}
	}

	@Ensure("increments_count_by_10")
	public void countTo10Indirectly() {
		final int old_count = count;

		for (int i = 0; i < 10; ++i) {
			invariant(lte(0, i) & lte(i, 10));
			invariant(eq(count, old_count + i));
			increment();
		}
	}

}

File: byteback-test/test-scala/src/main/scala/byteback/test/complete/Counter.scala
package byteback.test.complete;

import byteback.annotations.Contract._;
import byteback.annotations.Special._;
import byteback.annotations.Operator._;
import byteback.annotations.Operator.{eq => equal};

class Counter() {
  var count: Int = 0;

  @Predicate
  def increments_count_by_1(): Boolean = {
		return equal(count, old(count) + 1);
  }

  @Predicate
  def increments_count_by_10(): Boolean = {
		return equal(count, old(count) + 10);
  }

  @Ensure("increments_count_by_1")
  def increment() : Unit = {
    count = count + 1;
  }

  @Ensure("increments_count_by_10")
  def incrementTo10(): Unit = {
    var i: Int = 0;
		var old_count: Int = count;

    while (i < 10) {
			invariant(lte(0, i) & lte(i, 10));
			invariant(eq(count, old_count + i));

      increment();
      i = i + 1;
    }
  }

}
